Option for stored data in `jf job list`

FabiPi3 opened the pull request:
Add the option to display data from the `stored_data` field in the `Response` object. A User can select a key or multiple keys to display those values.

On a sidenote, another option was added to not display the uuid of the jobs. I typically don't gain much from this information since the db_id already gives me the unique information I need to find the job in the database.

Things to discuss:

- Name of the cli options (including short-cuts)
- Name of the column in the output table: For now only the key. Should be sufficient since the user does know what fields he/she requested, right?
- Name of the default_value which is displayed once no data is available or the specified key is not present. For now a light grey italic `none`

Example:

See the attached screenshot.

<img width="1215" alt="Bildschirmfoto 2024-12-10 um 11 32 32" src="https://github.com/user-attachments/assets/98a6bba2-cf8b-4485-b61e-5e3c707deecf">

Maybe closes #227

Commit:
option for stored data

gpetretto commented:
Hi @FabiPi3,

sorry for the delay, but I did not have much time to dedicate to jobflow-remote in the last days. I think the option for the `stored_data` is a good update.
Concerning the job id, I think it would be good to introduce a better customization of the displayed result, rather than just an option to skip one value. Otherwise we could end up with a very large number of options to control each displayed fields. I think it would be interesting to introduce a `--format`/`--output` option, to control the columns printed. With the additional option to directly customize this in the `~/.jfremote.yaml` configuration file. What do you think about this solution? And would you be interested/available to try implementing it? In that case I can give some more tips about the code that needs to be modified. No pressure anyway in case you would rather not work more on this point.

In any case, for all the features that will be added I will also ask you to implement a test in https://github.com/Matgenix/jobflow-remote/blob/develop/tests/db/cli/test_job.py before merging.

FabiPi3 commented:
I think you should upgrade the minimal typer version. I have troubles running the tests with typer 0.9.0, but newer versions does work. In the CI for python 3.12 typer version 0.12.something is used.

Commit:
implement the option to specify all columns explicitly

FabiPi3 commented:
I have added an option to explicitly specify all columns that will be displayed. This can also be set in the config file. I had to rewrite the formatting table function to make it more dynamic. Let me know what you think.

I have also added a small test, would that be sufficient?

I was just wondering about to how to specify the columns in the terminal. Right now I use the list option from typer, which can be quite lengthy to type: `jf -fe job list -m 3 -hk db_id -hk name -hk state -hk worker -hk locked -hk last_updated -hk run_time`

Maybe is would be better to just pass a string: `jf job list -hk db_id&name&state&worker` and than split this string afterward. What do you think?

gpetretto commented on src/jobflow_remote/cli/formatting.py:
```diff
@@ -26,79 +26,100 @@
-    if verbosity >= 2:
-        table.add_column("Lock id")
-        table.add_column("Lock time" + time_zone_str)
+    default_stored_data_value = Text.from_markup("[italic grey58]none[/]")
```
Why not leaving it just empty, as for the other fields that may not be filled?

FabiPi3 replied:
Ok

gpetretto commented on src/jobflow_remote/cli/job.py:
```diff
@@ -105,9 +106,25 @@ def jobs_list(
+        Optional[list[str]],
+        typer.Option(
+            "--header-keys",
+            "-hk",
```
I would use `--output` and `-o` as options, I think it is more intuitive and matches other commands like `df`, `lsblk`. What do you think?

FabiPi3 replied:
I am not familiar with those other commands but will stick to your suggestion here

gpetretto commented on src/jobflow_remote/cli/job.py:
```diff
@@ -105,9 +106,25 @@ def jobs_list(
+            "--header-keys",
+            "-hk",
+            help="Table columns to be shown. Overrides verbosity option. Can also be set in the config file",
+        ),
```
Maybe this should also contain the list of available options? Otherwise one would need to either look at the source code or proceed by trial and error. Would it be too long?

FabiPi3 replied:
Added it. I think its fine.

gpetretto commented on src/jobflow_remote/config/settings.py:
```diff
@@ -30,6 +30,10 @@ class JobflowRemoteSettings(BaseSettings):
         LogLevel.WARN, description="The level set for logging in the CLI"
     )
+    cli_job_list_columns: list[str] = Field(
+        default_factory=list,
```
I think it would be less ambiguous to have this as `Optional[list[str]]` and have `None` as a default, instead of an empty list.

FabiPi3 replied:
Ok.

gpetretto commented on src/jobflow_remote/config/settings.py:
```diff
@@ -30,6 +30,10 @@ class JobflowRemoteSettings(BaseSettings):
     )
+    cli_job_list_columns: list[str] = Field(
+        default_factory=list,
+        description="The list of columns to show in the `jf job list` command.",
```
Also here maybe it would be useful to have the full list?

FabiPi3 replied:
Actually I would like to avoid hard-coding those values. But importing them will give a circular import error. So either I need to move the map definition or it could point to the help page of the list command. What do you think?

gpetretto replied:
I think that referring to the help of `jf job list` for the list of available options should be enough.

FabiPi3 replied:
Great.

gpetretto commented on src/jobflow_remote/cli/job.py:
```diff
@@ -128,6 +145,11 @@ def jobs_list(
             worker_name,
         ],
     )
+    header_keys = header_keys or SETTINGS.cli_job_list_columns
```
There are two improvements that can be done here:
1) it would be good if the user is directly warned that passing header keys and verbose is incompatible. I think you can use the `check_incompatible_opt` function and just pass the two.
2) as things are now, if you set the defaults in the config file and then pass `-v`, at the moment it will be just ignored. If you want to see more columns you will be forced to pass it in CLI or modify the config file. So if there is a `-v` and `cli_job_list_columns` is set, the `-v` should have the priority.

FabiPi3 replied:
Done.

gpetretto reviewed (commented):
Hi @FabiPi3,

thanks a lot for implementing the additional options! I think the implementation is fine. I left some minor comments on the code. 
I agree with you that repeating the option multiple times is really inconvenient for this kind of field. I think it would be good to switch to a single string, as you suggested. But I would use a `,` as a separator, rather than a `&`. It seems more standard and usually easier to type on the keyboard.

The test is fine, but can you also add one where it fails by passing an option that does not exist? You can pass `error=True` to `run_check_cli` and check that (part of) the error message is present in the output. 
Also, if you add the error for incompatible options and that -v overrides the configuration, it would be good to have tests for those use cases as well.

Commit:
implement Guido's feedback

FabiPi3 commented:
Hi Guido,

thanks for your feedback. I tried implementing all of it. I have also changed my code a bit.

I added a unit to the run time field as it felt necessary to me. What do you think?

Regarding the tests, I added two failing tests. I am not sure how to modify the SETTINGS object for the test to test this. Do you have a standard procedure?

gpetretto commented:
Thanks for the updates. 
For the test with changing the `SETTINGS` you can do it as in this test: https://github.com/Matgenix/jobflow-remote/blob/41bf0b287e7fd60615222b2ef38d01ebdd1bfc19/tests/db/cli/test_project.py#L15

Aside from this and adjusting the description of `cli_job_list_columns`, I think it will be ready to be merged.
I see that the tests are failing, but it seems unrelated to your changes and it is probably due to an inaccurate tests based on the dates of the job report.

Commit:
add referring to help

Commit:
add test for settings

FabiPi3 commented:
Added your comments. I agree, the failing tests seemed unrelated to my changes.

gpetretto commented:
Thanks!
